src/v/storage/segment_deduplication_utils.cc

```diff
@@ -102,16 +102,23 @@ ss::future<model::offset> build_offset_map(
         if (cfg.asrc) {
             cfg.asrc->check();
         }
```

Comment:
> This
could lead to compaction racing with prefix truncation, and the
compaction being unable to read the segment's compacted index because it
was removed. This adds an appropriate read lock.

Curious if this merely avoids an error that occurs when the race happens, or if it actually allows compaction or truncation to complete--serializing one before/after the other?

When the race occurs (independent of locking, etc...) is compaction cancelled and restarted later? I had kind of assumed that this is what we did and would continue to do.

Reply:
>Curious if this merely avoids an error that occurs when the race happens

Just this one. I contemplated doing something more heavy-handed like locking all segments for the entire duration of compaction, but opted to keep it simple and rely on existing status quo for now (which means if the segment is closed during compaction, we throw).

>When the race occurs (independent of locking, etc...) is compaction cancelled and restarted later?

That's right. The current mechanism isn't super elegant in that we throw and rely on housekeeping to try again next time. I opened https://github.com/redpanda-data/redpanda/issues/14947 to track cleaning this up